domain/storage/service/storage.go

```diff
@@ -0,0 +1,176 @@
+		// TODO(axw) implement support for volumes.
+		return "", errors.Errorf("storage kind %q not supported", arg.Kind.String()).Add(coreerrors.NotSupported)
+	}
+	if !internalstorage.IsValidPoolName(arg.Pool) {
+		return "", errors.Errorf("pool name %q not valid", arg.Pool).Add(storageerrors.InvalidPoolNameError)
+	}
+	if err := arg.StorageName.Validate(); err != nil {
+		return "", errors.Capture(err)
+	}
+
+	poolDetails, err := s.st.GetStoragePoolByName(ctx, arg.Pool)
+	if errors.Is(err, storageerrors.PoolNotFoundError) {
+		poolDetails = storage.StoragePoolDetails{
+			Name:     arg.Pool,
+			Provider: arg.Pool,
+			Attrs:    map[string]string{},
+		}
+	} else if err != nil {
+		return "", errors.Capture(err)
+	}
+
+	var attr map[string]any
+	if len(poolDetails.Attrs) > 0 {
+		attr = transform.Map(poolDetails.Attrs, func(k, v string) (string, any) { return k, v })
+	}
+	cfg, err := internalstorage.NewConfig(poolDetails.Name, internalstorage.ProviderType(poolDetails.Provider), attr)
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+	filesystemInfo, err := s.importStorageFromProvider(ctx, cfg, arg.ProviderId)
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+
+	return s.st.ImportFilesystem(ctx, arg.StorageName, *filesystemInfo)
+}
+
+func (s *StorageService) importStorageFromProvider(ctx context.Context, cfg *internalstorage.Config, providerID string) (*storage.FilesystemInfo, error) {
+	registry, err := s.registryGetter.GetStorageRegistry(ctx)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	provider, err := registry.StorageProvider(cfg.Provider())
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	details, err := s.st.GetModelDetails()
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	resourceTags := map[string]string{
+		tags.JujuModel:      details.ModelUUID,
+		tags.JujuController: details.ControllerUUID,
+	}
+
+	// If the storage provider supports filesystems, import the filesystem,
+	// otherwise import a volume which will back a filesystem.
+	var filesystemInfo *storage.FilesystemInfo
+	callCtx := envcontext.WithoutCredentialInvalidator(ctx)
+	if provider.Supports(internalstorage.StorageKindFilesystem) {
+		filesystemInfo, err = s.importFilesystemFromProvider(callCtx, provider, cfg, providerID, resourceTags)
+	} else {
+		filesystemInfo, err = s.importVolumeFromProvider(callCtx, provider, cfg, providerID, resourceTags)
+	}
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	return filesystemInfo, nil
+}
+
+func (s *StorageService) importFilesystemFromProvider(ctx envcontext.ProviderCallContext, provider internalstorage.Provider, cfg *internalstorage.Config, providerID string, resourceTags map[string]string) (*storage.FilesystemInfo, error) {
+	filesystemSource, err := provider.FilesystemSource(cfg)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	filesystemImporter, ok := filesystemSource.(internalstorage.FilesystemImporter)
```

Comment:
There is a todo that exists on this interface to clean up this casting smell. Can we do it now and just add a func to the provider or source?

Reply:
fair. The func comment was to wrap this up like we do with environproviders and the provider tracker where you get a getter that you can request an interface off of. 

Removes all of this if then that from the structure and make the func very procedural. Just a nice to have.

Reply:
Trouble is the interface here is returned by calling a getter on the provider, and the capability of the result is dependent on the type of storage we are using. It will be a lot of work if we want to look at changing the implementation. Given this PR is primarily concerned with moving existing logic down from the facade layer to the service layer, let's keep it at that and look at provider implementation concerns separately.